### PR TITLE
[new release] cohttp, cohttp-top, cohttp-mirage, cohttp-lwt, cohttp-lwt-unix, cohttp-lwt-jsoo and cohttp-async (5.0.0)

### PR DESCRIPTION
**Pull request by mseri**
An OCaml library for HTTP clients and servers

- Project page: <a href="https://github.com/mirage/ocaml-cohttp">https://github.com/mirage/ocaml-cohttp</a>
- Documentation: <a href="https://mirage.github.io/ocaml-cohttp/">https://mirage.github.io/ocaml-cohttp/</a>

##### CHANGES:

- Cohttp.Header: new implementation (lyrm mirage/ocaml-cohttp#747)

  + New implementation of Header modules using an associative list instead of a map, with one major semantic change (function ```get```, see below), and some new functions (```clean_dup```, ```get_multi_concat```)
  + More Alcotest tests as well as fuzzing tests for this particular module.

  ### Purpose

  The new header implementation uses an associative list instead of a map to represent headers and is focused on predictability and intuitivity: except for some specific and documented functions, the headers are always kept in transmission order, which makes debugging easier and is also important for [RFC7230§3.2.2](https://tools.ietf.org/html/rfc7230#section-3.2.2) that states that multiple values of a header must be kept in order.

  Also, to get an intuitive function behaviour, no extra work to enforce RFCs is done by the basic functions. For example, RFC7230§3.2.2 requires that a sender does not send multiple values for a non list-value header. This particular rule could require the ```Header.add``` function to remove previous values of non-list-value headers, which means some changes of the headers would be out of control of the user. With the current implementation, an user has to actively call dedicated functions to enforce such RFCs (here ```Header.clean_dup```).

  ### Semantic changes
  Two functions have a semantic change : ```get``` and ```update```.

  #### get
  ```get``` was previously doing more than just returns the value associated to a key; it was also checking if the searched header could have multiple values: if not, the last value associated to the header was returned; otherwise, all the associated values were concatenated and returned. This semantics does not match the global idea behind the new header implementation, and would also be very inefficient.

  + The new ```get``` function only returns the last value associated to the searched header.
  + ```get_multi_concat``` function has been added to get a result similar to the previous ```get``` function.

  #### update
  ```update``` is a pretty new function (mirage/ocaml-cohttp#703) and changes are minor and related to ```get``` semantic changes.

  + ```update h k f``` is now modifying only the last occurrences of the header ```k``` instead of all its occurrences.
  + a new function ```update_all``` function has been added and work on all the occurrences of the updated header.

  ### New functions :

  + ```clean_dup```  enables the user to clean headers that follows the {{:https://tools.ietf.org/html/rfc7230#section-3.2.2} RFC7230§3.2.2} (no duplicate, except ```set-cookie```)
  + ```get_multi_concat``` has been added to get a result similar to the previous ```get``` function.

- Cohttp.Header: performance improvement (mseri, anuragsoni mirage/ocaml-cohttp#778)
  **Breaking** the headers are no-longer lowercased when parsed, the headers key comparison is case insensitive instead.

- cohttp-lwt-unix: Adopt ocaml-conduit 5.0.0 (smorimoto mirage/ocaml-cohttp#787)
  **Breaking** `Conduit_lwt_unix.connect`'s `ctx` param type chaged from `ctx` to  `ctx Lazy.t`

- cohttp-mirage: fix deprecated fmt usage (tmcgilchrist mirage/ocaml-cohttp#783)
- lwt_jsoo: Use logs for the warnings and document it (mseri mirage/ocaml-cohttp#776)
- lwt: Use logs to warn users about leaked bodies and document it (mseri mirage/ocaml-cohttp#771)
- lwt, lwt_unix: Improve use of logs and the documentation, fix bug in the Debug.enable_debug function (mseri mirage/ocaml-cohttp#772)
- lwt_jsoo: Fix exception on connection errors in chrome (mefyl mirage/ocaml-cohttp#761)
- lwt_jsoo: Fix `Lwt.wakeup_exn` `Invalid_arg` exception when a js
  stack overflow happens in the XHR completion handler (mefyl mirage/ocaml-cohttp#762).
- lwt_jsoo: Add test suite (mefyl mirage/ocaml-cohttp#764).

### CODE REVIEW

File: packages/cohttp-mirage/cohttp-mirage.5.0.0/opam [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+Please see <https://mirage.io> for a self-hosted explanation
+and instructions on how to use this library."""
+depends: [
+  "ocaml" {>= "4.08"}
+  "dune" {>= "2.0"}
+  "mirage-flow" {>= "2.0.0"}
+  "mirage-channel" {>= "4.0.0"}
+  "conduit" {>= "2.0.2"}
+  "conduit-mirage" {>= "2.0.2"}
```
Comment (mseri) [inline packages/cohttp-mirage/cohttp-mirage.5.0.0/opam:26]:
```suggestion
  "conduit-mirage" {>= "2.3.0"}
```

Uses Resolver_mirage.S and Conduit_mirage.Flow.S

**Comment by mseri** [comment]
This should be merged only once it is properly tested in the CI, and right now the CI is overloaded

**Comment by mseri** [comment]
`ez_api` conflicts were incorrect, fixed in https://github.com/ocaml/opam-repository/pull/20282
`h1_parser` was also fixed in https://github.com/ocaml/opam-repository/pull/20276

The rest is fine, this can be merged

**Comment by kit-ty-kate** [comment]
The constraints on zeit and h1_parser do not seem to be right. The same errors happen and they are still marked as compatible with this release

**Comment by mseri** [comment]
I think that was run with the incorrect fix, let me restart the CI

**Comment by kit-ty-kate** [comment]
It’s not the CI, i already restarted it twice

**Comment by mseri** [comment]
Oh, then it is quite suprising. I don't think there were breaking changes in the Headers for quite a while

**Comment by mseri** [comment]
The CI is installing cohttp 5.0.0 even if there is a conflict for cohttp >= 5

**Comment by kit-ty-kate** [comment]
Sadly there isn’t. See my comment on https://github.com/ocaml/opam-repository/pull/20276 for example

**Comment by mseri** [comment]
I have just seen your comments... I hope those constraints will be recognized in the future.
What do you think? Should I disable the tests if `cohttp:version >= 5.0.0`, or we just leave them there as is until they will work.

**Comment by kit-ty-kate** [comment]
I think it’s easier to disable them entirely (by simply removing the `with-test` part) for now and see later to unlock them for cohttp 5

**Comment by mseri** [comment]
I think this can be merged. Zeit can be fixed separately

**Pull request merged**